jekyll/_posts/2021-07-21-Nim-Efficienct-Expressive-Elecgant-Benchmarking.md

```diff
@@ -0,0 +1,369 @@
+      while `cullaid` < `szbitsid`:        
+        let `cptr0id` = cast[ptr uint8](`ca` + (`cullaid` shr 3))
+        `cptr0id`[] = `cptr0id`[] or cBITMSK[`cullaid` and 7]
+        `cullaid` += `bpintid`
+    csstmnt.add nnkOfBranch.newTree(
+      newLit(n),
+      ofbrstmnts
+    )
+  for n in 0x40'u8 .. 255'u8: # fill in defaults for remaining possibilities
+    csstmnt.add nnkOfBranch.newTree(
+      newLit(n),
+      nnkStmtList.newTree(
+        nnkBreakStmt.newTree(
+          newEmptyNode()
+        )
+      )
+    )
+  result.add quote do:
+    let `endalmtid` = `cmpstsalmtid` - `strt7id`
+    var `cullaid` = `ca` + `strt0id`
+    `csstmnt`
+#  echo csstmnt[9].astGenRepr # see AST for a given case
+#  echo csstmnt[9].toStrLit # see code for a given case
+#  echo result.toStrLit # see entire produced code at compile time
+
+proc benchSoE(): iterator(): int {.closure.} =
+  var cmpsts = newSeq[byte](16384)
+  let cmpstsa = cast[int](cmpsts[0].addr)
+  for _ in 0 ..< cLOOPS:
+    for i in 0 .. 254: # cull to square root of limit
+      if (cmpsts[i shr 3] and cBITMSK[i and 7]) == 0'u8: # if prime -> cull its composites
+        let bp = i +% i +% 3
+        let swi = (i +% i) *% (i +% 3) +% 3
+        unrollLoops(cmpstsa, 16384, swi, bp)
+  return iterator(): int {.closure.} =
+    yield 2 # the only even prime
+    for i in 0 .. 131071: # separate iteration over results
+      if (cmpsts[i shr 3] and cBITMSK[i and 7]) == 0'u8:
+        yield i +% i +% 3
+```
+Running the code with the above substitution in the same non-checking release mode as before results in a run time that is about 2.5 times faster than the previous version and over ten times faster than the naive implementation, meaning that the culling operation time is this ratio faster per operation.  For a modern CPU, one can calculate that it uses just over one CPU clock cycle per cull operation (198273 culling operations per loop).
+
+Now, this speed can't quite be maintained for a real application with larger ranges and thus larger base prime values as there will come a point where the overheads of the loop unrolling overcomes its benefit, but that is for quite large ranges of about a hundred million; as well, the loss above that point increases slowly since the base prime values only grow as the square root of the entire sieved range and also the base primes density becomes more sparse with increasing range such that using a page segmented implementation to count the primes up to ten's of billions only increases the time per culling operation to about two CPU clock cycles each, meaning that the average is still much less than pure "bit-twiddling".
+
+## Analysis as compared to other languages
+
+So in order to be able to assert that Nim is "Efficient, Effective, and Elegant", one needs to compare it to something; since Nim's design goals are in that order, languages that aren't as efficient will be rejected from the outset.  This means the following:
```

Comment:
It's "efficient, **expressive**, elegant" ;-)

Reply:
Dang, typing too fast in the rush to post ;-)

I don't know how to edit a file name in a PR.  Should I just close this PR, and I can edit in your suggested addition as to comparisons with dynamic languages as well as the file name change?